core/include/drivers/gpio.h

```diff
@@ -40,4 +66,126 @@ struct gpio_ops {
 			      enum gpio_interrupt ena_dis);
 };
 
+struct gpio {
+	struct gpio_chip *chip;
+	uint32_t dt_flags;
+	unsigned int pin;
+};
+
+static inline bool gpio_ops_is_valid(const struct gpio_ops *ops)
+{
+	return ops->set_direction && ops->get_direction && ops->get_value &&
+	       ops->set_value;
+}
+
+static inline void gpio_set_direction(struct gpio *gpio, enum gpio_dir dir)
+{
+	gpio->chip->ops->set_direction(gpio->chip, gpio->pin, dir);
+}
+
+static inline enum gpio_dir gpio_get_direction(struct gpio *gpio)
+{
+	return gpio->chip->ops->get_direction(gpio->chip, gpio->pin);
+}
+
+static inline void gpio_set_value(struct gpio *gpio, enum gpio_level value)
+{
+	if (gpio->dt_flags & GPIO_ACTIVE_LOW)
+		value = !value;
+
+	gpio->chip->ops->set_value(gpio->chip, gpio->pin, value);
+}
+
+static inline enum gpio_level gpio_get_value(struct gpio *gpio)
+{
+	enum gpio_level value = GPIO_LEVEL_LOW;
+
+	value = gpio->chip->ops->get_value(gpio->chip, gpio->pin);
+
+	if (gpio->dt_flags & GPIO_ACTIVE_LOW)
+		value = !value;
+
+	return value;
+}
+
+#if defined(CFG_DT) && defined(CFG_DRIVERS_GPIO)
+
+struct gpio *gpio_get_dt_two_cells(struct dt_driver_phandle_args *a,
+				   TEE_Result *res);
+
+/**
+ * gpio_dt_get_by_index - Get a gpio controller at a specific index in
+ * 'gpios' property
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of the subnode containing a 'gpios' property
+ * @index: Reset controller index in '*-gpios' property
```

Comment:
s/Reset controller index/GPIO pin index/

`gpio_name` is missing .

Reply:
Done